stan/math/laplace/laplace_marginal.hpp

```diff
@@ -0,0 +1,275 @@
+   * Newton step are stored and reused. 
+   */
+  struct laplace_marginal_density_vari : public vari {
+    /* dimension of the global parameters. */
+    int phi_size_;
+    /* global parameters. */
+    vari** phi_;
+    /* dimension of the latent Gaussian parameters. */
+    int theta_size_;
+    /* the marginal density of the observation, conditional on the 
+     * globl parameters. */
+    vari** marginal_density_;
+    /* An object to store the sensitivities of phi. */
+    Eigen::VectorXd phi_adj_;
+
+    template <typename T, typename K, typename D>
+    laplace_marginal_density_vari
+      (double marginal_density,
+       const Eigen::Matrix<T, Eigen::Dynamic, 1>& phi,
+       const std::vector<Eigen::VectorXd>& x,
+       const D& diff_likelihood,
+       const K& covariance_function,
+       const Eigen::VectorXd& theta,
+       const Eigen::VectorXd& W_root,
+       const Eigen::MatrixXd& L,
+       const Eigen::VectorXd& a,
+       const Eigen::VectorXd& l_grad)
+      : vari(marginal_density),
+        phi_size_(phi.size()),
+        phi_(ChainableStack::instance().memalloc_.alloc_array<vari*>(
+            phi_size_)),
+        theta_size_(theta.size()),
+        marginal_density_(
+          ChainableStack::instance().memalloc_.alloc_array<vari*>(1)) {
+      for (int i = 0; i < phi_size_; i++) phi_[i] = phi(i).vi_;
+
+      // CHECK -- is there a cleaner way of doing this?
+      marginal_density_[0] = this;
+      marginal_density_[0] = new vari(marginal_density, false);
+
+      // compute derivatives of covariance matrix with respect to phi.
+      covariance_sensitivities<K> f(x, theta_size_, covariance_function);
+      Eigen::MatrixXd diff_cov;
+      Eigen::MatrixXd covariance;
+      {
+        Eigen::VectorXd covariance_vector;
+        jacobian_fwd(f, value_of(phi), covariance_vector, diff_cov);
+        covariance = to_matrix(covariance_vector, theta_size_, theta_size_);
+      }
+
+      // Now compute the full gradient (using algorithm 5.1 of R & W)
+      // Check: is there an efficient way to solve / divide a diagonal matrix?
+      Eigen::MatrixXd Z;
+      {
+        Eigen::MatrixXd W_root_diag = W_root.asDiagonal();
+        Z = W_root_diag *
+              L.transpose().triangularView<Eigen::Upper>()
+               .solve(L.triangularView<Eigen::Lower>()
+                 .solve(W_root_diag));
+      }
+
+      Eigen::MatrixXd
+        C = mdivide_left_tri<Eigen::Lower>(L,
+                  diag_pre_multiply(W_root, covariance));
+
+      // CHECK -- should there be a minus sign here?
+      Eigen::VectorXd s2 = 0.5 * (covariance.diagonal()
+                 - (C.transpose() * C).diagonal())
+                 .cwiseProduct(diff_likelihood.third_diff(theta));
+
+      phi_adj_ = Eigen::VectorXd(phi_size_);
+      for (int j = 0; j < phi_size_; j++) {
+        Eigen::VectorXd j_col = diff_cov.col(j);
+        C = to_matrix(j_col, theta_size_, theta_size_);
```

Comment:
Unless you are 100% certain the C you defined above will always have `theta_size_` rows and `theta_size_` columns I'd be careful here

Reply:
I'm 100% sure.